federation/src/main/scala/caliban/federation/EntityResolver.scala

```diff
@@ -20,10 +21,42 @@ object EntityResolver {
     resolver: A => ZQuery[R, CalibanError, Option[T]]
   )(implicit schema: Schema[R, T]): EntityResolver[R] =
     new EntityResolver[R] {
-      override def resolve(value: InputValue): ZQuery[R, CalibanError, Step[R]] =
-        ZQuery.fromEither(implicitly[ArgBuilder[A]].build(value)).flatMap { arg =>
-          resolver(arg).map(_.fold[Step[R]](Step.NullStep)(schema.resolve))
-        }
+      override def resolve(value: InputValue): Step[R] =
+        QueryStep(
+          ZQuery.fromEither(implicitly[ArgBuilder[A]].build(value)).flatMap { arg =>
+            resolver(arg).map(_.fold[Step[R]](Step.NullStep)(schema.resolve))
+          }
+        )
+
+      override def toType: __Type = schema.toType_()
+    }
+
+  def fromFunction[A: ArgBuilder, T](
+    resolver: A => Option[T]
+  )(implicit schema: Schema[Any, T]): EntityResolver[Any] =
+    new EntityResolver[Any] {
+      override def resolve(value: InputValue): Step[Any] =
+        QueryStep(
+          ZQuery.fromEither(implicitly[ArgBuilder[A]].build(value)).map { arg =>
+            resolver(arg).fold[Step[Any]](Step.NullStep)(schema.resolve)
+          }
+        )
```

Comment:
We can avoid wrapping into an effect altogether
```suggestion
        ArgBuilder[A].build(value) match {
          case Right(arg)  =>
            resolver(arg) match {
              case Some(value) => schema.resolve(value)
              case _           => Step.NullStep
            }
          case Left(error) => Step.FailureStep(error)
        }
```